### ISSUE TITLE
Is CheckTx in recheck mode run in parallel with Proposal ? (Tendermint 0.34.11)

### ISSUE THREAD

**Issue by kaldubin** [tendermint/tendermint#8350]
Hello Tendermint!
Here is a issue/maybe a bug, that we encountered recently. We managed to avoid it with some cheap trick but if what we understood from it is real, we think it might be useful to present it in the documentation for awareness.

**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source): 0.34.11

**ABCI app** (name for built-in, URL for self-written if it's publicly available): Our app runs federative learning (FL) tasks. The project is recent so the code is not as its finest though here is how it should work. 
The Validators Nodes create a block containing a Model and a Request (`select`) for 4 Light Clients to perform some task and send back their Results. 
Once the 4 results are processed into the blockchain a new Model is created by each Validators, so we have 4 new Model but they are identical and we want to put only one of those into the blockchain.
And so the process goes on and on: with each new Model is a new Request of 4 clients, once the 4 Results are received and compute we produce a new Model with a new Request etc.

**Environment**:
- **OS** (e.g. from /etc/os-release): Ubuntu 20.04
- **Install tools**: py-abci
- **Others**:

**What happened**: Running our application, at some point we observe two identical new Model being put in two separate blocks. The second one did not passed the **CheckTx** transaction in its node though and should not be in the blockchain.

```
Checking transaction
Considering transaction : 2 - -3 - 3f073419f8b2c97a08f95d0772887e6729a02558f50ea21c5c204f71234c5c0e
BEGIN-BLOCK
DeliverTX: tx = 2, -2, 3f073419f8b2c97a08f95d0772887e6729a02558f50ea21c5c204f71234c5c0e
valid transaction
END-BLOCK
BEGIN COMMIT
============== BLOCK BEGINS =============
=============== NEW MODEL ===============
round: 2 - score: 0.098 - select: [0, 5, 1, 7]
=============== BLOCK ENDS ==============
END COMMIT
Checking transaction
NOT Considering transaction: 2 - -3 - 3f073419f8b2c97a08f95d0772887e6729a02558f50ea21c5c204f71234c5c0e
	Reason: already commited
BEGIN-BLOCK
DeliverTX: tx = 2, -3, 3f073419f8b2c97a08f95d0772887e6729a02558f50ea21c5c204f71234c5c0e
Invalid transaction (NEW_MODEL not in context)
END-BLOCK
BEGIN COMMIT
============== BLOCK BEGINS =============
=============== NEW MODEL ===============
round: 2 - score: 0.098 - select: [6, 1, 5, 2]
=============== BLOCK ENDS ==============
END COMMIT
```
Here we can see the transaction `{2 | -3 | 3f073419f8...}` did not passed the **CheckTx** (message "NOT Considering transaction ...") but is delivered in the new block (as invalid) anyway. Note that this transaction was already in the Mempool (already checked in the previous step as we can see above), so it is a **CheckTx** in **recheck mode**.
Looking at the logs from the node being proposer at this height we can see this:
```
[39;49;22mI[2022-04-14|11:50:28.108] committed state                              module=state height=16 num_txs=1 app_hash=65653061623666336565336434656233373537623631373830396561383531613266323234356664626663353136633936633537383438396231646666643265
[30;1mD[2022-04-14|11:50:28.110] recheck txs                                  module=mempool numtxs=1 height=16
[39;49;22m[30;1mD[2022-04-14|11:50:28.116] entering new round                           module=consensus height=17 round=0 current=17/0/RoundStepNewHeight
[39;49;22m[30;1mD[2022-04-14|11:50:28.117] entering propose step                        module=consensus height=17 round=0 current=17/0/RoundStepNewRound
[39;49;22m[30;1mD[2022-04-14|11:50:28.117] node is a validator                          module=consensus height=17 round=0
[39;49;22m[30;1mD[2022-04-14|11:50:28.117] propose step; our turn to propose            module=consensus height=17 round=0 proposer=40EEE39C1001737CBA135A396D55450C1D35CC4E
[39;49;22m[30;1mD[2022-04-14|11:50:28.117] Received tick                                module=consensus old_ti="{Duration:906.06908ms Height:17 Round:0 Step:RoundStepNewHeight}" new_ti="{Duration:8s Height:17 Round:0 Step:RoundStepPropose}"
[39;49;22m[30;1mD[2022-04-14|11:50:28.117] Scheduled timeout                            module=consensus dur=8s height=17 round=0 step=RoundStepPropose
[39;49;22mI[2022-04-14|11:50:28.120] indexed block                                module=txindex height=16
[39;49;22m[30;1mD[2022-04-14|11:50:28.129] indexed block txs                            module=txindex height=16 num_txs=1
[39;49;22m[30;1mD[2022-04-14|11:50:28.129] setHasVote                                   module=consensus peerH/R=16/0 H/R=16/0 type=SIGNED_MSG_TYPE_PRECOMMIT index=2
[39;49;22m[30;1mD[2022-04-14|11:50:28.132] signed proposal                              module=consensus height=17 round=0 proposal="Proposal{17/0 (35DB17D34903434E53181D901B57D104A182D590F53897091FDEC8878D692129:5:20973954784C, -1) 674841AA2F26 @ 2022-04-14T11:50:28.129215516Z}"
[39;49;22m[30;1mD[2022-04-14|11:50:28.133] Sending vote message                         module=consensus ps="PeerState{\n  Key        439fbf42cec4a0f64b2c452a778ef72a81066594\n  RoundState PeerRoundState{\n    17/0/RoundStepNewHeight @2022-04-14 11:50:28.116052224 +0000 UTC\n    Proposal 0:000000000000 -> nil-BitArray\n    POL      BA{4:____} (round -1)\n    Prevotes   BA{4:____}\n    Precommits BA{4:____}\n    LastCommit BA{4:_xxx} (round 0)\n    Catchup    BA{4:____} (round -1)\n  }\n  Stats      peerStateStats{votes: 31, blockParts: 14}\n}" vote="Vote{0:40EEE39C1001 16/00/SIGNED_MSG_TYPE_PRECOMMIT(Precommit) F3BF24FD4650 B825660B98C5 @ 2022-04-14T11:50:28.009530837Z}"
[39;49;22m[30;1mD[2022-04-14|11:50:28.133] setHasVote                                   module=consensus peerH/R=17/0 H/R=16/0 type=SIGNED_MSG_TYPE_PRECOMMIT index=0
[39;49;22m[30;1mD[2022-04-14|11:50:28.133] Picked rs.LastCommit to send                 module=consensus peer="Peer{MConn{172.17.0.5:35328} 439fbf42cec4a0f64b2c452a778ef72a81066594 in}" height=17
[39;49;22m[30;1mD[2022-04-14|11:50:28.133] adding vote                                  module=consensus vote_height=16 vote_type=SIGNED_MSG_TYPE_PRECOMMIT val_index=3 cs_height=17
[39;49;22mI[2022-04-14|11:50:28.136] received proposal                            module=consensus proposal="Proposal{17/0 (35DB17D34903434E53181D901B57D104A182D590F53897091FDEC8878D692129:5:20973954784C, -1) 674841AA2F26 @ 2022-04-14T11:50:28.129215516Z}"
[39;49;22mI[2022-04-14|11:50:28.154] received complete proposal block             module=consensus height=17 hash=35DB17D34903434E53181D901B57D104A182D590F53897091FDEC8878D692129
[39;49;22m[30;1mD[2022-04-14|11:50:28.154] entering prevote step                        module=consensus height=17 round=0 current=17/0/RoundStepPropose
[39;49;22m[30;1mD[2022-04-14|11:50:28.155] prevote step: ProposalBlock is valid         module=consensus height=17 round=0
[39;49;22m[30;1mD[2022-04-14|11:50:28.175] signed and pushed vote                       module=consensus height=17 round=0 vote="Vote{0:40EEE39C1001 17/00/SIGNED_MSG_TYPE_PREVOTE(Prevote) 35DB17D34903 1A0822DDAF98 @ 2022-04-14T11:50:29.009530837Z}"
[39;49;22m[30;1mD[2022-04-14|11:50:28.177] adding vote                                  module=consensus vote_height=17 vote_type=SIGNED_MSG_TYPE_PREVOTE val_index=0 cs_height=17
[39;49;22m[30;1mD[2022-04-14|11:50:28.178] added vote to prevote                        module=consensus vote="Vote{0:40EEE39C1001 17/00/SIGNED_MSG_TYPE_PREVOTE(Prevote) 35DB17D34903 1A0822DDAF98 @ 2022-04-14T11:50:29.009530837Z}" prevotes="VoteSet{H:17 R:0 T:SIGNED_MSG_TYPE_PREVOTE +2/3:<nil>(0.25) BA{4:x___} map[]}"
[39;49;22m[30;1mD[2022-04-14|11:50:28.181] tx is no longer valid                        module=mempool tx=5EC211F09BC7DE4B82A62C9B2A56593FC5554424E5AC9FB14B5E2793D5E95CFA res="&{CheckTx:code:1 }" err=null
[39;49;22m[30;1mD[2022-04-14|11:50:28.183] done rechecking txs                          module=mempool 
```
(Though some messages have been remove for clarity, sorry for this huge logs)
The **CheckTx** in **recheck mode** seems to start right after the last block is committed (normal) but is done after the **ProposalBlock** is sent. I am not completely familiar with Tendermint but it seems the ProposalBlock included the transaction as it is marked as valid in the mempool (from the previous step), even though the **recheck** process was not finished. As if the **CheckTx** in **recheck mode** was ran concurrently with the **Propose** process by Tendermint.

**What you expected to happen**:
As the transaction did not passed the **CheckTx** (in recheck mode), we expected it to be evicted from the mempool and not considered for the next block.

**Have you tried the latest version**: no

**How to reproduce it** (as minimally and precisely as possible):
To be honest, we don't really know why it is acting like that although we found out some necessary parameters for it to happen.
- `skip_timeout_commit = true`
- `create_empty_blocks = false` 
- And the app_hash computed in the commit function change when a new model is added and validate
- Maybe heavy transactions ?

**Config (you can paste only the changes you've made)**:
```
skip_timeout_commit = true
create_empty_blocks = false
cache_size = 0
max_body_bytes = 20000000
max_tx_bytes = 20000000
laddr = "tcp://0.0.0.0:26657"
addr_book_strict = false
pex = false
timeout_propose = "8s"
timeout_prevote = "6s"
timeout_precommit = "6s"
```

**node command runtime flags**:
```
--p2p.persistent_peers "..."
--moniker node_x
--home "test_scenario/tm_home"
```

**Anything else we need to know**:
We changed `skip_timeout_commit` to false (as default) and the thing seems to work for now but if the Propose process and the Recheck process are indeed ran concurrently then this issue might appear again.
We also saw the issue #2384 and the proposal #7898 so it seems this issue will be corrected in the next release with ABCI++. But we did not see duplicates of such issue and thought it would be useful to present it here.
Thanks!
